ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,450 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
 
 static
 NTSTATUS
-OpenRegistryHandlesFromSymbolicLink(IN PUNICODE_STRING SymbolicLinkName,
-                                    IN ACCESS_MASK DesiredAccess,
-                                    IN OPTIONAL PHANDLE GuidKey,
-                                    IN OPTIONAL PHANDLE DeviceKey,
-                                    IN OPTIONAL PHANDLE InstanceKey)
+IopBuildSymbolicLink(IN PUNICODE_STRING PrefixString,
+                     IN PUNICODE_STRING DeviceString,
+                     IN PUNICODE_STRING GuidString,
+                     IN OPTIONAL PUNICODE_STRING ReferenceString,
+                     OUT PUNICODE_STRING SymbolicLinkName)
 {
-    OBJECT_ATTRIBUTES ObjectAttributes;
-    UNICODE_STRING BaseKeyU;
-    UNICODE_STRING GuidString, SubKeyName, ReferenceString;
-    PWCHAR StartPosition, EndPosition;
-    HANDLE ClassesKey;
-    PHANDLE GuidKeyRealP, DeviceKeyRealP, InstanceKeyRealP;
-    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    UNICODE_STRING SymbolicLink;
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    /* Build up new symolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = PrefixString->Length + DeviceString->Length + sizeof(L"#") + GuidString->Length + ReferenceString->Length;
+    SymbolicLink.Buffer = ExAllocatePoolWithTag(NonPagedPool, SymbolicLink.MaximumLength, TAG_IO);
+    if (!SymbolicLink.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        ExFreePoolWithTag(SymbolicLink.Buffer, TAG_IO);
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
 
-    if (GuidKey != NULL)
-        GuidKeyRealP = GuidKey;
-    else
-        GuidKeyRealP = &GuidKeyReal;
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, PrefixString);
+    if (!NT_SUCCESS(Status))
```

Comment:
You can just `NT_ASSERT(NT_SUCCESS(Status))` (or `NT_VERIFY(NT_SUCCESS(RtlAppend...))`) on all of the append operations, since it should be impossible for them to fail if your math above is correct. No need for the if/dprint/return.

Reply:
Done.